Add beam eigenemittances to reduced diagnostics.

cemitch99 opened the pull request:
This PR adds calculation of the 3 eigenemittances (or mode emittances) of the 6D beam distribution from particle data as an additional reduced diagnostic.  Addresses Issue #663.  

- [x] implement auxiliary math routines needed to evaluate the characteristic polynomial and find its roots
- [x] implement calculation of eigenemittances from a covariance matrix
- [x] add additional required moment calculations in ReducedBeamCharacteristics.cpp
  - [x] skip eigenemittance columns from CSV output if not enabled
  - [x] add normalized emittance independently (always)
- [x] make reduced diagnostic calculation optional, and add print to output
- [x] add numerical benchmark example
- [x] add ParmParse statement to enable/disable calculation
- [x] check numerical precision for special degenerate cases (zero discriminant)
- [x] add documentation
   - [x] document user-facing inputs/python control
- [x] fix pytest using test_transformation.py
- [x] add liteature citations

cemitch99 committed:
Add eigenemittance calculation.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update CovarianceMatrixMath.H

Clean Doxygen and delete unused function TestMult (used for benchmarking).

cemitch99 committed:
Update EmittanceInvariants.cpp

Clean Doxygen

cemitch99 committed:
Update EmittanceInvariants.H

Clean Doxygen

cemitch99 committed:
Update EmittanceInvariants.cpp

Remove duplicate Doxygen comments in EmittanceInvariants.cpp/h

cemitch99 committed:
Treat special case of vanishing discriminant in Cardano's formula.

cemitch99 committed:
Test from ReducedDiagnostics; update ej ordering.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Add diagnostic output.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 commented on src/particles/diagnostics/CovarianceMatrixMath.H:
```diff
@@ -0,0 +1,137 @@
+
+        }
+
+        std::cout << "Discriminant, Q, R " << discriminant << " " << Q << " " << R << "\n";
+        std::cout << "Return x1, x2, x3 " << x1 << " " << x2 << " " << x3 << "\n";
```
```suggestion
        //std::cout << "Discriminant, Q, R " << discriminant << " " << Q << " " << R << "\n";
        //std::cout << "Return x1, x2, x3 " << x1 << " " << x2 << " " << x3 << "\n";
```

cemitch99 commented on src/particles/diagnostics/EmittanceInvariants.cpp:
```diff
@@ -0,0 +1,138 @@
+        I2 = -impactx::diagnostics::TraceMat(S2)/2.0_prt;
+        I4 = +impactx::diagnostics::TraceMat(S4)/2.0_prt;
+        I6 = -impactx::diagnostics::TraceMat(S6)/2.0_prt;
+
+        std::cout << "Return I2, I4, I6 " << I2 << " " << I4 << " " << I6 << "\n";
```
```suggestion

        //std::cout << "Return I2, I4, I6 " << I2 << " " << I4 << " " << I6 << "\n";
```

cemitch99 commented on src/particles/diagnostics/EmittanceInvariants.cpp:
```diff
@@ -0,0 +1,138 @@
+        amrex::ParticleReal d = -pow(I2,3)/6.0_prt + I2*I4/2.0_prt - I6/3.0_prt;
+
+        // Return the cubic coefficients
+        std::cout << "Return a,b,c,d " << a << " " << b << " " << c << " " << d << "\n";
```
```suggestion
        //std::cout << "Return a,b,c,d " << a << " " << b << " " << c << " " << d << "\n";
```

cemitch99 commented on src/particles/diagnostics/ReducedBeamCharacteristics.cpp:
```diff
@@ -322,6 +390,42 @@ namespace impactx::diagnostics
+        Sigma(6,1) = 1.0;
+
+        std::tuple <amrex::ParticleReal,amrex::ParticleReal,amrex::ParticleReal> emittances = Eigenemittances(Sigma);
+        std::cout << "Emittances = " << std::get<0>(emittances) << " " << std::get<1>(emittances) << " " << std::get<2>(emittances);
```
```suggestion
        //std::cout << "Emittances = " << std::get<0>(emittances) << " " << std::get<1>(emittances) << " " << std::get<2>(emittances);
```

cemitch99 committed:
Comment print statements used for debugging.

cemitch99 committed:
Update ReducedBeamCharacteristics.cpp

Comment benchmarking call to Eigenemittance to avoid unused variable errors.

cemitch99 commented:
This works, but suffers a loss of precision in certain cases (eg, uncoupled beams with et << ex = ey) due to the appearance of a small discriminant in Cardano's formula.  A fix for this is under investigation.

Some tests failing for unrelated reasons due to pytest.

cemitch99 committed:
Add benchmark example.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Add documentation.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update CMakeLists.txt

Corrected path to benchmark test files.

cemitch99 committed:
Update CMakeLists.txt

Update path for analysis scripts.

cemitch99 committed:
Normalize momenta by mc (dynamical units) for eigenemittances.

cemitch99 committed:
Add alternative routine for cubic roots.

cemitch99 commented on src/particles/diagnostics/EmittanceInvariants.cpp:
```diff
@@ -0,0 +1,138 @@
+        // Caution: The order of e1,e2,e3 should be consistent with the
+        // order ex,ey,et in the limit of uncoupled transport.
+        // The order below is important and has been checked.
+        roots = CubicRootsAlg(a,b,c,d);
```
There are two options for this function call - `CubicRootsAlg` and `CubicRootsTrig`.  The results should be equivalent, modulo differences in numerical precision.  This addresses the preceding comment.

cemitch99 committed:
Update EmittanceInvariants.cpp

Update default cubic roots alg.

cemitch99 committed:
Update CovarianceMatrixMath.H

Avoid subtracting complex and double types.

cemitch99 committed:
Temporarily comment pytest test_transformation.

cemitch99 commented on tests/python/test_transformation.py:
```diff
@@ -81,7 +81,7 @@ def test_transformation():
         if not np.isclose(val, rbc_s[key], rtol=rtol, atol=atol):
             print(f"initial[{key}]={val}, final[{key}]={rbc_s[key]} not equal")
-        assert np.isclose(val, rbc_s[key], rtol=rtol, atol=atol)
+//        assert np.isclose(val, rbc_s[key], rtol=rtol, atol=atol)
```
This change is temporary, so that we can validate the current PR.

cemitch99 replied:
Change removed.

cemitch99 committed:
Update test_transformation.py

Modify comment syntax c++ -> Python.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 commented on src/particles/diagnostics/EmittanceInvariants.cpp:
```diff
@@ -0,0 +1,138 @@
+        I4 = +impactx::diagnostics::TraceMat(S4)/2.0_prt;
+        I6 = -impactx::diagnostics::TraceMat(S6)/2.0_prt;
+
+        //std::cout << "Return I2, I4, I6 " << I2 << " " << I4 << " " << I6 << "\n";
```
```suggestion
```

cemitch99 commented on src/particles/diagnostics/CovarianceMatrixMath.H:
```diff
@@ -0,0 +1,211 @@
+
+        }
+
+        //std::cout << "Discriminant, Q, R " << discriminant << " " << Q << " " << R << "\n";
+        //std::cout << "Return x1, x2, x3 " << x1 << " " << x2 << " " << x3 << "\n";
```
```suggestion
```

cemitch99 commented on src/particles/diagnostics/CovarianceMatrixMath.H:
```diff
@@ -0,0 +1,211 @@
+
+        //std::cout << "Discriminant, Q, R " << discriminant << " " << Q << " " << R << "\n";
+        //std::cout << "Re(C), Im(C)  " << C.m_real << " " << C.m_imag << "\n";
+        //std::cout << "Return x1, x2, x3 " << x1 << " " << x2 << " " << x3 << "\n";
```
```suggestion
```

cemitch99 committed:
Delete commented debugging print statements.

cemitch99 committed:
Update EmittanceInvariants.H

Add literature citations.

cemitch99 committed:
Update EmittanceInvariants.cpp

Add literature citations.

cemitch99 commented on src/particles/diagnostics/EmittanceInvariants.cpp:
```diff
@@ -0,0 +1,141 @@
+        // Construct the coefficients of the cubic polynomial.
+        // This expression for the characteristic polynomial can be found in:
+        // V. Balandin, W. Decking, and N. Golubeva, "Relations Between Projected
+        // Emittances and Eigenemittances," in IPAC2018, Shanghai, China, 2013,
```
```suggestion
        // Emittances and Eigenemittances," in IPAC2013, Shanghai, China, 2013,
```

cemitch99 commented on src/particles/diagnostics/EmittanceInvariants.cpp:
```diff
@@ -0,0 +1,141 @@
+        // order ex,ey,et in the limit of uncoupled transport.
+        // The order below is important and has been checked.
+        roots = CubicRootsTrig(a,b,c,d);
+        amrex::ParticleReal e1 = sqrt(std::get<1>(roots));
+        amrex::ParticleReal e2 = sqrt(std::get<2>(roots));
+        amrex::ParticleReal e3 = sqrt(std::get<0>(roots));
```
```suggestion
        amrex::ParticleReal e1 = sqrt(abs(std::get<1>(roots)));
        amrex::ParticleReal e2 = sqrt(abs(std::get<2>(roots)));
        amrex::ParticleReal e3 = sqrt(abs(std::get<0>(roots)));
```

cemitch99 committed:
Apply suggestions from code review

Correct typo; avoid possible NaN resulting from sqrt.

cemitch99 commented on src/particles/diagnostics/EmittanceInvariants.cpp:
```diff
@@ -0,0 +1,141 @@
+        // order ex,ey,et in the limit of uncoupled transport.
+        // The order below is important and has been checked.
+        roots = CubicRootsTrig(a,b,c,d);
+        amrex::ParticleReal e1 = sqrt(abs(std::get<1>(roots)));
+        amrex::ParticleReal e2 = sqrt(abs(std::get<2>(roots)));
+        amrex::ParticleReal e3 = sqrt(abs(std::get<0>(roots)));
```
```suggestion
        amrex::ParticleReal e1 = sqrt(amrex::abs(std::get<1>(roots)));
        amrex::ParticleReal e2 = sqrt(amrex::abs(std::get<2>(roots)));
        amrex::ParticleReal e3 = sqrt(amrex::abs(std::get<0>(roots)));
```

cemitch99 committed:
Update src/particles/diagnostics/EmittanceInvariants.cpp

Use amrex::abs

cemitch99 committed:
Update EmittanceInvariants.cpp

Revert to std

cemitch99 committed:
Add emittance_1,2,3 print.

cemitch99 committed:
Finalize print of emittance_1,2,3.

cemitch99 committed:
Add ParmParse for optional calculation.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update test_transformation.py

Re-enable test_transformation.py.

ax3l commented on src/particles/diagnostics/CovarianceMatrixMath.H:
```diff
@@ -0,0 +1,206 @@
+#include <AMReX_REAL.H>
+#include <AMReX_GpuComplex.H>
+#include <tuple>
+#include <cmath>
```
```suggestion
#include <ablastr/constant.H>

#include <AMReX_Array.H>
#include <AMReX_REAL.H>
#include <AMReX_GpuComplex.H>

#include <cmath>
#include <tuple>

```

ax3l commented on src/particles/diagnostics/CovarianceMatrixMath.H:
```diff
@@ -0,0 +1,206 @@
+        amrex::ParticleReal x2 = 0.0_prt;
+        amrex::ParticleReal x3 = 0.0_prt;
+
+        amrex::ParticleReal Q = (3.0_prt*a*c - pow(b,2))/(9.0_prt * pow(a,2));
```
Please add `std::` before all `pow` to use the C++ instead of the C math functions.

ax3l commented on src/particles/diagnostics/CovarianceMatrixMath.H:
```diff
@@ -0,0 +1,206 @@
+        } else {
+
+           //Three real roots in trigonometric form.
+           amrex::ParticleReal theta = acos(R/sqrt(-pow(Q,3)));
```
```suggestion
           amrex::ParticleReal theta = std::acos(R/::sqrt(-::pow(Q,3)));
```

and below

ax3l commented on src/particles/diagnostics/EmittanceInvariants.H:
```diff
@@ -0,0 +1,75 @@
+#include <AMReX_REAL.H>
+
+#include <string>
+#include <unordered_map>
```
```suggestion
#include <AMReX_Array.H>
#include <AMReX_REAL.H>

#include <string>

```

ax3l commented on src/particles/diagnostics/EmittanceInvariants.cpp:
```diff
@@ -0,0 +1,141 @@
+#include <AMReX_GpuComplex.H>
+#include <cmath>
+#include <vector>
+#include <tuple>
```
```suggestion
#include <AMReX_Extension.H>
#include <AMReX_REAL.H>
#include <AMReX_GpuComplex.H>

#include <cmath>
#include <stdexcept>
#include <tuple>
#include <vector>
```

ax3l commented on src/particles/diagnostics/EmittanceInvariants.cpp:
```diff
@@ -0,0 +1,141 @@
+        // a change of sign.
+        for (int i = 1; i < 7; i++) {
+            for (int j = 1; j < 7; j++) {
+                if (j % 2) {
```
To be more explicit and avoiding an implicit int-to-bool cast, maybe:
```suggestion
                if (j % 2 == 1) {
```

ax3l commented on examples/coupled_optics/README.rst:
```diff
@@ -0,0 +1,51 @@
+.. _examples-coupled-optics:
+
+Coupled optics
+================
```
```suggestion
Coupled Optics
==============
```

ax3l commented on src/particles/diagnostics/EmittanceInvariants.cpp:
```diff
@@ -0,0 +1,141 @@
+        amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> Sigma
+    )
+    {
+        using namespace amrex::literals;
```
Let's add this to profiling:
```suggestion
        BL_PROFILE("impactx::diagnostics::Eigenemittances");

        using namespace amrex::literals;
```

ax3l reviewed (commented):
Thank you!!

I took a quick first look and added first inline suggestions. More thorough tomorrow :)

cemitch99 committed:
Apply suggestions from code review

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

cemitch99 committed:
Update EmittanceInvariants.cpp

Add PROFILE header.

cemitch99 committed:
Update CovarianceMatrixMath.H

Use math functions from std::  (sqrt, pow, sin, cos, etc) -> (std::sqrt, std::pow, std::sin, std::cos, etc).

cemitch99 committed:
Update src/particles/diagnostics/CovarianceMatrixMath.H

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

cemitch99 committed:
Update CovarianceMatrixMath.H

cemitch99 committed:
Update CovarianceMatrixMath.H

Remove std::math for complex functions.

cemitch99 commented on src/particles/diagnostics/CovarianceMatrixMath.H:
```diff
@@ -0,0 +1,209 @@
+        Complex Qc(Q,0.0_prt);
+        Complex Rc(R,0.0_prt);
+        Complex Dc(discriminant,0.0_prt);
+        Complex C = pow(-Rc + sqrt(Dc),1.0/3.0);
```
```suggestion
        Complex C = amrex::pow(-Rc + amrex::sqrt(Dc),1.0/3.0);
```

cemitch99 commented on src/particles/diagnostics/CovarianceMatrixMath.H:
```diff
@@ -0,0 +1,209 @@
+
+           Complex z1 = Qc/C - C;
+           Complex z2 = Qc/(xi*C) - xi*C;
+           Complex z3 = Qc/(pow(xi,2)*C) - pow(xi,2)*C;
```
```suggestion
           Complex z3 = Qc/(amrex::pow(xi,2)*C) - amrex::pow(xi,2)*C;
```

cemitch99 committed:
Use amrex:: math for complex functions.

ax3l commented on examples/CMakeLists.txt:
```diff
@@ -959,3 +959,19 @@ add_impactx_test(dogleg.py
+)
+add_impactx_test(coupled-optics.py
+    examples/coupled_optics/run_coupled_optics.py
+      OFF  # ImpactX MPI-parallel
```
```suggestion
    ON   # ImpactX MPI-parallel
    examples/coupled_optics/analysis_coupled_optics.py
    OFF  # no plot script yet
)
add_impactx_test(coupled-optics.py
    examples/coupled_optics/run_coupled_optics.py
    OFF  # ImpactX MPI-parallel
```

ax3l commented on src/particles/diagnostics/CovarianceMatrixMath.H:
```diff
@@ -0,0 +1,209 @@
+    std::tuple<
+            amrex::ParticleReal,
+            amrex::ParticleReal,
+            amrex::ParticleReal>
```
```suggestion
    std::tuple<
        amrex::ParticleReal,
        amrex::ParticleReal,
        amrex::ParticleReal
    >
```

ax3l commented on src/particles/diagnostics/CovarianceMatrixMath.H:
```diff
@@ -0,0 +1,209 @@
+        using namespace amrex::literals;
+        using ablastr::constant::math::pi;
+
+        std::tuple <amrex::ParticleReal,amrex::ParticleReal,amrex::ParticleReal> roots;
```
```suggestion
        std::tuple<amrex::ParticleReal, amrex::ParticleReal, amrex::ParticleReal> roots;
```

ax3l commented on src/particles/diagnostics/CovarianceMatrixMath.H:
```diff
@@ -0,0 +1,209 @@
+        amrex::ParticleReal tol = 1.0e-12;  //allow for roundoff error
+        if(discriminant > tol){
+
+           std::cout << "Polynomial in CubicRoots has one or more complex roots." << "\n";
```
This would fit well for the error channel:
```suggestion
           std::cerr << "impactx::diagnostics::CubicRootsTrig Polynomial has one or more complex roots." << "\n";
```

ax3l replied:
Actually, even better, is to use our warning manager:
- include:
https://github.com/ECP-WarpX/impactx/blob/4a7da60f72bb5149dca9d23924ee16dca7421d46/src/initialization/InitMeshRefinement.cpp#L15
- usage example:
https://github.com/ECP-WarpX/impactx/blob/4a7da60f72bb5149dca9d23924ee16dca7421d46/src/initialization/InitMeshRefinement.cpp#L51-L59

This one handles MPI well and groups warnings into user-friendly sections of the output.

cemitch99 replied:
Ok, done.

ax3l commented on src/particles/diagnostics/CovarianceMatrixMath.H:
```diff
@@ -0,0 +1,209 @@
+     * @param[in] d coefficient of constant term
+     * @returns tuple of three real roots
+     */
+    std::tuple<
+            amrex::ParticleReal,
+            amrex::ParticleReal,
+            amrex::ParticleReal>
```
```suggestion
    std::tuple<
        amrex::ParticleReal,
        amrex::ParticleReal,
        amrex::ParticleReal
    >
```

ax3l commented on src/particles/diagnostics/CovarianceMatrixMath.H:
```diff
@@ -0,0 +1,209 @@
+        using namespace amrex::literals;
+        using Complex = amrex::GpuComplex<amrex::ParticleReal>;
+
+        std::tuple <amrex::ParticleReal,amrex::ParticleReal,amrex::ParticleReal> roots;
```
```suggestion
        std::tuple<amrex::ParticleReal, amrex::ParticleReal, amrex::ParticleReal> roots;
```

ax3l commented on src/particles/diagnostics/CovarianceMatrixMath.H:
```diff
@@ -0,0 +1,209 @@
+           x3 = z3.m_real - b/(3.0*a);
+        }
+
+
```
```suggestion
```

ax3l commented on src/particles/diagnostics/CovarianceMatrixMath.H:
```diff
@@ -0,0 +1,209 @@
+
+        }
+
+
```
```suggestion
```

ax3l commented on src/particles/diagnostics/CovarianceMatrixMath.H:
```diff
@@ -0,0 +1,209 @@
+
+        // Discriminant should be < 0.  Otherwise, keep theta at default and throw an error.
+        amrex::ParticleReal tol = 1.0e-12;  //allow for roundoff error
+        if(discriminant > tol){
```
```suggestion
        if (discriminant > tol) {
```

ax3l commented on src/particles/diagnostics/CovarianceMatrixMath.H:
```diff
@@ -0,0 +1,209 @@
+
+           std::cout << "Polynomial in CubicRoots has one or more complex roots." << "\n";
+
+        } else if (Q == 0.0_prt){  // Special case of a triple root
```
```suggestion
        } else if (Q == 0.0_prt) {  // Special case of a triple root
```

ax3l commented on src/particles/diagnostics/CovarianceMatrixMath.H:
```diff
@@ -0,0 +1,209 @@
+
+        //Three roots in algebraic form.
+
+        if (C.m_real == 0.0_prt && C.m_imag == 0.0_prt){  // Special case of a triple root
```
```suggestion
        if (C.m_real == 0.0_prt && C.m_imag == 0.0_prt) {  // Special case of a triple root
```

ax3l commented on src/particles/diagnostics/CovarianceMatrixMath.H:
```diff
@@ -0,0 +1,209 @@
+     */
+    amrex::ParticleReal
+    TraceMat (
+        amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> A
```
We can avoid to copy the matrix A:
```suggestion
        amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> const & A
```

ax3l commented on src/particles/diagnostics/CovarianceMatrixMath.H:
```diff
@@ -0,0 +1,209 @@
+     */
+    amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6>
+    MultiplyMat (
+        amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> A,
+        amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> B
```
Avoid copying A and B:
```suggestion
        amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> const & A,
        amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> const & B
```

ax3l commented on src/particles/diagnostics/EmittanceInvariants.H:
```diff
@@ -0,0 +1,76 @@
+     * @param[in] Sigma symmetric 6x6 covariance matrix
+     * @returns tuple containing invariants I2, I4, and I6
+     */
+    std::tuple<
+            amrex::ParticleReal,
+            amrex::ParticleReal,
+            amrex::ParticleReal>
```
```suggestion
    std::tuple<
        amrex::ParticleReal,
        amrex::ParticleReal,
        amrex::ParticleReal
    >
```

ax3l commented on src/particles/diagnostics/EmittanceInvariants.H:
```diff
@@ -0,0 +1,76 @@
+            amrex::ParticleReal,
+            amrex::ParticleReal>
+    KineticInvariants (
+        amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> Sigma
```
```suggestion
        amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> const & Sigma
```

ax3l commented on src/particles/diagnostics/EmittanceInvariants.cpp:
```diff
@@ -0,0 +1,145 @@
+            amrex::ParticleReal,
+            amrex::ParticleReal>
+    KineticInvariants (
+        amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> Sigma
```
```suggestion
        amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> const & Sigma
```

ax3l commented on src/particles/diagnostics/EmittanceInvariants.cpp:
```diff
@@ -0,0 +1,145 @@
+     * input - Sigma symmetric 6x6 covariance matrix
+     * returns - tuple containing invarants I2, I4, and I6
+     */
+    std::tuple<
+            amrex::ParticleReal,
+            amrex::ParticleReal,
+            amrex::ParticleReal>
```
```suggestion
    std::tuple<
        amrex::ParticleReal,
        amrex::ParticleReal,
        amrex::ParticleReal
    >
```

ax3l commented on src/particles/diagnostics/EmittanceInvariants.cpp:
```diff
@@ -0,0 +1,145 @@
+            amrex::ParticleReal,
+            amrex::ParticleReal>
+    Eigenemittances (
+        amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> Sigma
```
```suggestion
        amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> const & Sigma
```

ax3l commented on src/particles/diagnostics/EmittanceInvariants.H:
```diff
@@ -0,0 +1,76 @@
+            amrex::ParticleReal,
+            amrex::ParticleReal>
+    Eigenemittances (
+        amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> Sigma
```
```suggestion
        amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> const & Sigma
```

ax3l commented on src/particles/diagnostics/EmittanceInvariants.cpp:
```diff
@@ -0,0 +1,145 @@
+
+        using namespace amrex::literals;
+
+        std::tuple <amrex::ParticleReal,amrex::ParticleReal,amrex::ParticleReal> invariants;
+        std::tuple <amrex::ParticleReal,amrex::ParticleReal,amrex::ParticleReal> roots;
+        std::tuple <amrex::ParticleReal,amrex::ParticleReal,amrex::ParticleReal> emittances;
```
```suggestion
        std::tuple<amrex::ParticleReal, amrex::ParticleReal, amrex::ParticleReal> invariants;
        std::tuple<amrex::ParticleReal, amrex::ParticleReal, amrex::ParticleReal> roots;
        std::tuple<amrex::ParticleReal, amrex::ParticleReal, amrex::ParticleReal> emittances;
```

ax3l commented on tests/python/test_transformation.py:
```diff
@@ -81,7 +81,7 @@ def test_transformation():
         if not np.isclose(val, rbc_s[key], rtol=rtol, atol=atol):
             print(f"initial[{key}]={val}, final[{key}]={rbc_s[key]} not equal")
-        assert np.isclose(val, rbc_s[key], rtol=rtol, atol=atol)
+            assert np.isclose(val, rbc_s[key], rtol=rtol, atol=atol)
```
Not sure this change is needed.
```suggestion
        assert np.isclose(val, rbc_s[key], rtol=rtol, atol=atol)
```

ax3l replied:
Alternatively, we can just skip doing the `isclose` twice:
```suggestion
            assert False
```

ax3l reviewed (commented):
This looks great!

ax3l committed:
Avoid copying matrices

ax3l committed:
Formatting: Whitespaces

Only spaces

ax3l commented on src/particles/diagnostics/EmittanceInvariants.cpp:
```diff
@@ -0,0 +1,145 @@
+        // a change of sign.
+        for (int i = 1; i < 7; i++) {
+            for (int j = 1; j < 7; j++) {
+                if (j % 2 == 1) {
```
More robust is:
```suggestion
                if (j % 2 != 0) {
```
but yes, we only do positive numbers here.

ax3l commented on src/particles/diagnostics/ReducedBeamCharacteristics.cpp:
```diff
@@ -322,6 +400,9 @@ namespace impactx::diagnostics
         data["dispersion_y"] = dispersion_y;
         data["dispersion_py"] = dispersion_py;
         data["charge_C"] = charge;
+        data["emittance_1"] = emittance_1;
```
As discussed in the developer meeting last Wed:
Instead of redefining the meaning of `emittance_N`, we rather would skip writing it if the diagnostics was disabled.

ax3l commented on docs/source/dataanalysis/dataanalysis.rst:
```diff
@@ -93,6 +93,8 @@ The code writes out the values in an ASCII file prefixed ``reduced_beam_characte
     Horizontal and vertical dispersion (unit: meter)
 * ``dispersion_px``, ``dispersion_py``
     Derivative of horizontal and vertical dispersion (unit: dimensionless)
+* ``emittance_1``, ``emittance_2``, ``emittance_3``
```
We need to update `usage/parameters.rst` and `usage/python.rst` to advertise the inputs control to calculate these.

Here, we might need to add one more line referencing these sections, so users know these entries are only added conditionally.

ax3l commented on src/particles/diagnostics/ReducedBeamCharacteristics.cpp:
```diff
@@ -283,6 +305,62 @@ namespace impactx::diagnostics
+
+        // Parse the diagnostic parameters
+        amrex::ParmParse pp_diag("diag");
+        bool compute_eigenemittances = false;
```
My feeling is we should enable this by default, but maybe there is a good reason not to.
```suggestion
        bool compute_eigenemittances = true;
```

cemitch99 replied:
The eigenemittance calculation is delicate for highly symmetric beams in uncoupled optics (where it is also not needed), and it causes one of the tests in pyTest to fail.  This, combined with its high complexity compared to other reduced diagnostics, suggests to me that it should be enabled if needed by the user.

ax3l reviewed (commented):
Avoid copies of matrices

ax3l committed:
More general odd-ness test

cemitch99 committed:
Add conditional emittance  output columns.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Add documentation for user-facing input control.

cemitch99 committed:
Update dataanalysis documentation of eigenemittances.

cemitch99 commented on docs/source/dataanalysis/dataanalysis.rst:
```diff
@@ -84,7 +84,7 @@ The code writes out the values in an ASCII file prefixed ``reduced_beam_characte
 * ``sig_px``, ``sig_py``, ``sig_pt``
     Standard deviation of the particle momentum deviations (energy difference for ``pt``) normalized by the magnitude of the reference particle momentum (unit: dimensionless)
 * ``emittance_x``, ``emittance_y``, ``emittance_t``
-    Normalized rms beam emittance (unit: meter)
```
This was documented incorrectly.

cemitch99 committed:
Update CovarianceMatrixMath.H

Update treatment of warnings.

cemitch99 committed:
Apply suggestions from code review

Changes to test_transformation.py were not intentional.  These were reversed per code review.

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

ax3l commented on tests/python/test_transformation.py:
```diff
@@ -81,7 +81,7 @@ def test_transformation():
         if not np.isclose(val, rbc_s[key], rtol=rtol, atol=atol):
             print(f"initial[{key}]={val}, final[{key}]={rbc_s[key]} not equal")
-            assert np.isclose(val, rbc_s[key], rtol=rtol, atol=atol)
+        assert np.isclose(val, rbc_s[key], rtol=rtol, atol=atol)
```
```suggestion
            assert False
```

ax3l committed:
Update tests/python/test_transformation.py

ax3l reviewed (approved):
Thank you, that looks great!
